addons/mail/wizard/mail_compose_message.py

```diff
@@ -359,91 +523,206 @@ def action_save_as_template(self):
-        if mass_mail_mode and self.model:
-            rendered_values = self.render_message(res_ids)
-        # compute alias-based reply-to in batch
-        reply_to_value = dict.fromkeys(res_ids, None)
-        if mass_mail_mode and not self.reply_to_force_new:
-            records = self.env[self.model].browse(res_ids)
-            reply_to_value = records._notify_get_reply_to(default=False)
-            # when having no specific reply-to, fetch rendered email_from value
-            for res_id, reply_to in reply_to_value.items():
-                if not reply_to:
-                    reply_to_value[res_id] = rendered_values.get(res_id, {}).get('email_from', False)
+        additional_values_all = {}
+        if rendering_mode and self.model:
+            additional_values_all = self._prepare_mail_values_dynamic(res_ids)
+        elif not rendering_mode:
+            additional_values_all = self._prepare_mail_values_static(res_ids)
 
         for res_id in res_ids:
-            # static wizard (mail.message) values
-            mail_values = {
+            additional_values = additional_values_all.get(res_id, {})
+            mail_values = dict(base_values, **additional_values)
+            mail_values_all[res_id] = mail_values
+
+        mail_values_all = self._process_state(mail_values_all)
+        return mail_values_all
+
+    def _prepare_mail_values_common(self):
+        """Prepare values always valid, not rendered or dynamic whatever the
+        composition mode and related records. """
+        self.ensure_one()
+        email_mode = self.composition_mode == 'mass_mail'
+
+        if email_mode:
+            subtype_id = False
+        elif self.subtype_id:
+            subtype_id = self.subtype_id.id
+        else:
+            subtype_id = self.env['ir.model.data']._xmlid_to_res_id('mail.mt_comment')
+
+        values = {
+            'author_id': self.author_id.id,
+            'mail_activity_type_id': self.mail_activity_type_id.id,
+            'message_type': 'email' if email_mode else self.message_type,
+            'parent_id': self.parent_id.id,
+            'reply_to_force_new': self.reply_to_force_new,
+            'scheduled_date': self.scheduled_date,
+            'subtype_id': subtype_id,
+        }
+        if self.composition_mode == 'comment':
+            # Several custom layouts make use of the model description at rendering, e.g. in the
+            # 'View <document>' button. Some models are used for different business concepts, such as
+            # 'purchase.order' which is used for a RFQ and and PO. To avoid confusion, we must use a
+            # different wording depending on the state of the object.
+            # Therefore, we can set the description in the context from the beginning to avoid falling
+            # back on the regular display_name retrieved in ``_notify_by_email_prepare_rendering_context()``.
+            model_description = self._context.get('model_description')
```

Comment:
```suggestion
            model_description = self.env.context.get('model_description')
```
nitpicky, but i find it clearer to use env.context instead of _context, it makes it more clear that the context depends on the env.